Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx

```diff
@@ -288,13 +400,163 @@ void vtkSlicerMarkupsWidgetRepresentation2D::UpdateAllPointsAndLabelsFromMRML(do
         continue;
         }
 
+      // get control point position
       double slicePos[3] = { 0.0 };
       this->GetNthControlPointDisplayPosition(pointIndex, slicePos);
-
       controlPoints->ControlPoints->InsertNextPoint(slicePos);
-      slicePos[0] += labelsOffset / sqrt(2.0);
-      slicePos[1] += labelsOffset / sqrt(2.0);
-      this->Renderer->SetDisplayPoint(slicePos);
+
+      // set normal label position
+      double labelPos[3] = { 0.0 };
+      labelPos[0] = slicePos[0] + labelsOffset / sqrt(2.0);
+      labelPos[1] = slicePos[1] + labelsOffset / sqrt(2.0);
+
+      std::string labelStr;
+      int labelLength = 0;
+
+      labelStr = markupsNode->GetNthControlPointLabel(pointIndex);
+
+      // make label texts multiline if <br> is used
+      std::string from = "<br>";
+      std::string to = "\n";
+      int start_pos = 0;
+      while ((start_pos = labelStr.find(from, start_pos)) != std::string::npos)
+        {
+        labelStr.replace(start_pos, from.length(), to);
+        start_pos += to.length();
+        }
+
+      // find length of longest string segment after splitting up into multiline segments
+      std::string s = labelStr;
+      std::string delimiter = "\n";
+      int longestStringLen = 0;
+      int numberLines = 1;
+
+      size_t pos = 0;
+      std::string token;
+      while ((pos = s.find(delimiter)) != std::string::npos)
+        {
+        numberLines++;
+        token = s.substr(0, pos);
+        if (token.length() > longestStringLen)
+          {
+          longestStringLen = token.length();
+          }
+        s.erase(0, pos + delimiter.length());
+        }
+      if (s.length() > longestStringLen)
+        {
+        longestStringLen = s.length();
+        }
+
+      if (this->MarkupsDisplayNode->GetPointLabelsDistanceScale() > 0.0 &&
+        markupsNode->GetNthControlPointPositionVisibility(pointIndex) &&
+          this->MarkupsDisplayNode->GetPointLabelsVisibility() &&
+                !markupsNode->GetNthControlPointLabel(pointIndex).empty()) // add not occluded
+        {
+        double lineStartPos[3] = { 0.0 };
+        double lineEndPos[3] = { 0.0 };
+        const int* screenSize = this->Renderer->GetRenderWindow()->GetSize();
+        int screenWidth = screenSize[0];
+        int screenHeight = screenSize[1];
+        labelLength = longestStringLen;
+
+        // Not working correctly - giving the exact same values in _bb  wherever you place this code
+        //double _bb[4] = { 0.0 };
+        //this->TextActor->GetBoundingBox(this->Renderer, _bb);
+        //this->TextActor->SetTextScaleModeToViewport();
+        double labelCharWidth =  this->MarkupsDisplayNode->GetTextScale() * 3.;
+        double labelTextWidth = labelCharWidth * labelLength;
```

Comment:
Getting the label size by counting characters is slow and inaccurate. It should be possible to get it from VTK. Check out how it is done in vtkLabelPlacementMapper or vtkDynamic2DLabelMapper. Maybe the best would be to fix vtkLabelPlacementMapper (that supports leader lines).